### PR TITLE
Change btsieve behaviour to be query driven

### LINKED ISSUE #1346: Finish mob-programming branch
Branch is here: https://github.com/comit-network/comit-rs/tree/btsieve-mob-programming

Tasks that need to be done:
- remove/resolve TODOs (can also create tickets for TODOs)
- use `BitcoindHttpBlockSource` (https://github.com/comit-network/comit-rs/issues/1131): this will need a small refactoring because of the API changes introduced in this PR (`BlockSource` is now called `MatchingTransactions`, takes the query and does the matching inside)
- Remove `BlockchainDag` and related structs & traits
- Probably remove the `MatchBlock` trait and inline the functionality into the respective implementation of `MatchingTransactions`

The tests _should_ pass after this, if so, create PR.

### PR DESCRIPTION
**Pull request by da-kami**
* poll for blocks for each query upon query creation
* blocksources implement query matching for transactions
* blocksources return matching transactions, but we still store the id
* remove bitcoin/ethereum Blockchain and BlockchainDag implementation
* remove block processors

fixes #1346

### CODE REVIEW

File: btsieve/src/ethereum/web3_http_blocksource.rs [hunk 1]
```diff
@@ -61,7 +71,70 @@ impl BlockSource for Web3HttpBlockSource {
...
+
+impl MatchingTransactions<EventQuery> for Web3HttpBlockSource {
+    type Error = Error;
+    type Transaction = ethereum_support::Transaction;
+
+    fn matching_transactions(
+        &self,
+        query: EventQuery,
+    ) -> Box<dyn Stream<Item = Self::Transaction, Error = Self::Error> + Send + 'static> {
+        let poll_interval = match self.network {
```
Comment (D4nte) [inline btsieve/src/ethereum/web3_http_blocksource.rs]: More relevant to the bitcoin one (can't find it) but I think it would be more readable to warp the match inside the `Duration::from_secs` so that it's clear these figures are seconds. (again, it's more for the bitcoin one which is every few hundred thousand seconds)

File: btsieve/src/ethereum/web3_http_blocksource.rs [hunk 2]
```diff
@@ -61,7 +71,70 @@ impl BlockSource for Web3HttpBlockSource {
...
+                                    log::debug!(target: "ethereum::blocksource", "Transport error encountered during polling: {:?}", e);
+                                    Ok(None)
+                                },
+                                _ => Err(error)
+                            }
+                        })
+                        .map_err(Error::Web3)
+                }
             })
             .filter_map(|maybe_block| maybe_block)
```
Comment (D4nte) [inline btsieve/src/ethereum/web3_http_blocksource.rs]: Why is that needed?

File: btsieve/src/bitcoin/bitcoind_http_blocksource.rs [hunk 3]
```diff
@@ -135,13 +137,24 @@ impl BlockSource for BitcoindHttpBlockSource {
...
-                            }
+                            },
+                            _ => Err(error)
                         }
                     })
-                    .map_err(blocksource::Error::Source)
             })
             .filter_map(|maybe_block| maybe_block);
 
+        let stream = stream
```
[the remark this answers is not in the document]
Comment (luckysori) [reply btsieve/src/bitcoin/bitcoind_http_blocksource.rs]: I don't mind the separation, but I would rename the first one to `blocks` and this one to `matching_transactions`.

File: btsieve/src/ethereum/web3_http_blocksource.rs [hunk 4]
```diff
@@ -61,7 +71,70 @@ impl BlockSource for Web3HttpBlockSource {
                             _ => Err(error)
                         }
                     })
-                    .map_err(Error::Source)
+                    .map_err(Error::Web3)
+            })
+            .filter_map(|maybe_block| maybe_block)
```
[the remark this answers is not in the document]
Comment (luckysori) [reply btsieve/src/ethereum/web3_http_blocksource.rs]: I think it's because the previous combinator returns `Result<Option<Block>,Error>`. This will take all the `Option<Block>`s and only yield the `Some(Block)`s, filtering out the `None`s.

File: btsieve/src/routes.rs [hunk 5]
```diff
@@ -162,14 +164,26 @@ pub fn delete_query<
...
+    QRR: QueryResultRepository<Q>,
+    C: 'static + Send + Sync,
+    E: 'static,
+    BS: MatchingTransactions<Q, Error = E>,
+>(
     _client: Arc<C>,
     _network: String,
-    query_repository: Arc<QR>,
+    blocksource: Arc<BS>,
+    active_query_repository: Arc<QR>,
```
Comment (luckysori) [inline btsieve/src/routes.rs]: :question: Why did the name change here? I don't mind adding the active, but I'd add it everywhere. Otherwise it can get confusing.

File: btsieve/src/routes.rs [hunk 6]
```diff
@@ -178,13 +192,26 @@ pub fn get_or_create_query<Q: Send + Eq, QR: QueryRepository<Q>, C: 'static + Se
...
+                    let future = blocksource
+                        .matching_transactions(query)
+                        .map_err(|_| ())
+                        .for_each(move |transaction| {
+                            query_result_repository
+                                .add_result(id.clone(), transaction.into_transaction_id());
+                            Ok(())
+                        });
+
+                    tokio::spawn(future);
```
Comment (luckysori) [inline btsieve/src/routes.rs]: :question: Should we be explicit about the runtime? Does it matter?
